batch/interface.go

```diff
@@ -0,0 +1,32 @@
+package batch
+
+import (
+	"github.com/coreos/bbolt"
+)
+
+// Request defines an operation that can be batched into a single bbolt
+// transaction.
+type Request struct {
+	// Update is applied alongside other operations in the batch.
+	//
+	// NOTE: This method MUST NOT acquire any mutexes.
+	Update func(tx *bbolt.Tx) error
+
+	// OnSuccess is applied if the batch or a subset of the batch including
+	// this request all succeeded without failure.
+	//
+	// NOTE: This field is optional.
+	OnSuccess func(err error) error
+
+	// Solo is applied if this request's Update failed, and was asked to
+	// retry outside of the batch.
+	Solo func() error
+}
+
+// Scheduler abstracts a generic batching engine that accumulates an incoming
+// set of Requests, executes them, and returns the error from the operation.
+type Scheduler interface {
+	// Execute schedules a Request for execution with the next available
```

Comment:
Wondering in case the time-based scheduler increases its duration, in which case the callers of this function may not want to block until it completes :thinking:

Reply:
good call, i've updated the godoc. yes, the latency on the commit could be higher than before. for the initial use case though we already execute all of our concurrent requests within goroutines (during channel graph validation)